google/cloud/security/scanner/audit/forwarding_rule_rules_engine.py

```diff
@@ -0,0 +1,249 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Rules engine for forwarding rules engine"""
+from collections import namedtuple
+
+from google.cloud.security.common.gcp_type.resource import ResourceType
+from google.cloud.security.common.util import log_util
+from google.cloud.security.scanner.audit import base_rules_engine as bre
+from google.cloud.security.scanner.audit import errors as audit_errors
+
+
+LOGGER = log_util.get_logger(__name__)
+
+class ForwardingRuleRulesEngine(bre.BaseRulesEngine):
+    """Rules engine for forwarding rules"""
+
+    RuleViolation = namedtuple('RuleViolation',
+                               [
+                                   'violation_type', 'target', 'rule_index',
+                                   'load_balancing_scheme', 'port_range',
+                                   'resource_type', 'port', 'ip_protocol',
+                                   'ip_address', 'resource_id'])
+
+    def __init__(self, rules_file_path, snapshot_timestamp=None):
+        """Initialize.
+
+        Args:
+            rules_file_path (str): file location of rules
+            snapshot_timestamp (str): snapshot timestamp. Defaults to None.
+                If set, this will be the snapshot timestamp
+                used in the engine.
+        """
+        super(ForwardingRuleRulesEngine, self).__init__(
+            rules_file_path=rules_file_path,
+            snapshot_timestamp=snapshot_timestamp)
+        self.rule_book = None
+
+    def build_rule_book(self, global_configs=None):
+        """Build forwarding rules rule book from the rules definition file.
+
+        Args:
+            global_configs (dict): Global configurations.
+        """
+        self.rule_book = ForwardingRuleRulesBook(self._load_rule_definitions())
+
+    # pylint: disable=arguments-differ
+    def find_policy_violations(self, forwarding_rule,
+                               force_rebuild=False):
+        """Determine whether forwarding rule violates rules.
+
+        Args:
+            forwarding_rule (ForwardingRule): The one specific forwarding rule
+                to be matched against all white list rules
+            force_rebuild (bool): If True, rebuilds the rule book. This will
+                reload the rules definition file and add the rules to the book.
+
+        Returns:
+            RuleViolation: A rule violation tuple with all data about the
+                forwarding rule that didnt pass a white list
+        """
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+        resource_rules = self.rule_book.get_resource_rules()
+        matched = False
+        for rule in resource_rules:
+            matched = rule.find_match(forwarding_rule) or matched
```

Comment:
I am wondering, do you need the "or matched" part, since you return on finding a match (next line)?